RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+
+  // Calculation of dR2 for Clustering
+  ALPAKA_FN_ACC ALPAKA_FN_INLINE static float dR2(Position4 pos1, Position4 pos2) {
+    float mag1 = sqrtf(pos1.x * pos1.x + pos1.y * pos1.y + pos1.z * pos1.z);
+    float cosTheta1 = mag1 > 0.0 ? pos1.z / mag1 : 1.0;
+    float eta1 = 0.5 * logf((1.0 + cosTheta1) / (1.0 - cosTheta1));
+    float phi1 = atan2f(pos1.y, pos1.x);
+
+    float mag2 = sqrtf(pos2.x * pos2.x + pos2.y * pos2.y + pos2.z * pos2.z);
+    float cosTheta2 = mag2 > 0.0 ? pos2.z / mag2 : 1.0;
+    float eta2 = 0.5 * logf((1.0 + cosTheta2) / (1.0 - cosTheta2));
+    float phi2 = atan2f(pos2.y, pos2.x);
+
+    float deta = eta2 - eta1;
+    float dphi = std::abs(std::abs(phi2 - phi1) - M_PI) - M_PI;
+    return (deta * deta + dphi * dphi);
+  }
+
+  // Get index of seed
+  ALPAKA_FN_ACC static auto getSeedRhIdx(int* seeds, int seedNum) { return seeds[seedNum]; }
+
+  // Get rechit fraction of a given rechit for a given seed
+  ALPAKA_FN_ACC static auto getRhFrac(reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+                                      int topoSeedBegin,
+                                      reco::PFRecHitFractionDeviceCollection::View fracView,
+                                      int seedNum,
+                                      int rhNum) {
+    int seedIdx = pfClusteringVars[topoSeedBegin + seedNum].topoSeedList();
+    return fracView[pfClusteringVars[seedIdx].seedFracOffsets() + rhNum].frac();
+  }
+
+  // Cluster position calculation
+  template <bool debug = false>
+  ALPAKA_FN_ACC static auto computeClusterPos(reco::PFClusterParamsDeviceCollection::ConstView pfClusParams,
+                                              Position4& pos4,
+                                              float frac,
+                                              int rhInd,
+                                              reco::PFRecHitDeviceCollection::ConstView pfRecHits,
+                                              float rhENormInv) {
+    Position4 rechitPos = Position4{pfRecHits[rhInd].x(), pfRecHits[rhInd].y(), pfRecHits[rhInd].z(), 1.0};
+    const auto rh_energy = pfRecHits[rhInd].energy() * frac;
+    const auto norm = (frac < pfClusParams.minFracInCalc() ? 0.0f : std::max(0.0f, logf(rh_energy * rhENormInv)));
+    if constexpr (debug)
+      printf("\t\t\trechit %d: norm = %f\tfrac = %f\trh_energy = %f\tpos = (%f, %f, %f)\n",
+             rhInd,
+             norm,
+             frac,
+             rh_energy,
+             rechitPos.x,
+             rechitPos.y,
+             rechitPos.z);
+
+    pos4.x += rechitPos.x * norm;
+    pos4.y += rechitPos.y * norm;
+    pos4.z += rechitPos.z * norm;
+    pos4.w += norm;  //  position_norm
+  }
+
+  // Processing single seed clusters
+  // Device function designed to be called by all threads of a given block
+  template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>, bool debug = false>
```

Comment:
This now needs to be called as `hcalFastCluster_singleSeed<TAcc, void, true>(...)` to enable debugging.

I would reorder the template parameters as
```suggestion
  template <bool debug = false, typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
```

This way one can enable debugging simply with `hcalFastCluster_singleSeed<true>(...)`.